src/nvim/lua/vim.lua

```diff
@@ -489,4 +489,73 @@ function vim.defer_fn(fn, timeout)
   return timer
 end
 
+local on_keystroke_callbacks = {}
+
+--- Register a lua {fn} with an {id} to be run after every keystroke.
+---
+--- NOTE: {fn} will be automatically removed if an error occurs while calling.
+---
+--@param ns_id number: See |nvim_create_namesapce()|
+--@param fn function: Function to call. It should take one argument, which is a string.
+--                    The string will contain the literal keys typed.
+--                    See |i_CTRL-V|
+function vim.register_keystroke_callback(ns_id, fn)
+  vim.validate {
+    fn = { fn, 'c', true},
```

Comment:
If this is optional to allow using register to *also* remove things, then we should remove the remove_keystroke_callback function and say that you can use `nil` to clear it. Otherwise the optionality should be documented.